### PR TITLE
Adds the ability to define Middleware that should be ran on a specific controller

### PR DESCRIPTION
**Pull request by ketzusaka**
Open to comments on this one. This particular feature is inspired by Rails before_filter and Laravel's Controller Middleware. It opens the doors for certain controllers to be behind a logged in user, while others open to everyone. This is a very common need in most web applications, and I'm sure there are plenty more usecases where you'd want a middleware to run on only a certain controller.

### CODE REVIEW

File: Sources/Vapor/Routing/Application+Route.swift [hunk 1]
```diff
@@ -64,13 +64,27 @@ extension Application {
...
         // ie: /users/:id
-        self.get(fullPath) { try controller.init().show($0) }
-        self.put(fullPath) { try controller.init().update($0) }
-        self.delete(fullPath) { try controller.init().destroy($0) }
+        self.add(.Get, path: fullPath, action: RoutedController.show)
+        self.add(.Put, path: fullPath, action: RoutedController.update)
+        self.add(.Delete, path: fullPath, action: RoutedController.destroy)
+    }
+
+    public final func add<RoutedController: Controller>(method: Request.Method, path: String, action: (RoutedController) -> (Request) throws -> ResponseConvertible) {
```
Comment (LoganWright) [inline Sources/Vapor/Routing/Application+Route.swift]: This seems like a good addition for the potential of additional custom controllers and it cleans up the declarations in post above.

File: Sources/Vapor/Routing/Application+Route.swift [hunk 2]
```diff
@@ -64,13 +64,27 @@ extension Application {
...
+        var handler: Request.Handler = { request in
+            let controller = RoutedController()
+            let actionCall = action(controller)
+            return try actionCall(request).response()
+        }
+
+        for mw in RoutedController.middleware {
```
Comment (LoganWright) [inline Sources/Vapor/Routing/Application+Route.swift]: I think we should stick with the original scoped middleware implementation and not bind controllers to this. I think we can achieve the same effect working within the framework

**Comment by ketzusaka** [comment]
Thanks. Going to close this and just add the add method. Didn't realize there was scoped grouping of middleware.

I do feel like there is value in a controller specifying middleware that it runs for itself, but I'll do that separately.

File: Sources/Vapor/Routing/Controller.swift [hunk 3]
```diff
@@ -2,4 +2,12 @@ import Foundation
 
 public protocol Controller {
     init()
-}
\ No newline at end of file
+    static var middleware: [Middleware.Type] { get }
```
Comment (LoganWright) [inline Sources/Vapor/Routing/Controller.swift]: I think we should leave this one to the scoped middleware on application for consistency, ie:

```Swift
app.middleware([scopedMiddlewares]) {
    // handle and add scoped controller here
}
```

**Pull request closed without merging**